### ISSUE TITLE
Get_Battery_Status Extended Message isn't functioning

### ISSUE THREAD

**Issue by davedesro** [STMicroelectronics/stm32-mw-usbpd-core#1]
# Overview
The library does not allow the developer to set / control the value of `Battery Status Ref` when calling **Get_Battery_Status** extended message

# Details
When calling 
```
uint8_t data = 0; 
USBPD_PE_SendExtendedMessage(0, USBPD_SOPTYPE_SOP, USBPD_EXT_GET_BATTERY_STATUS, &data, 1);
HAL_Delay(300); /**<-- for simplicity, to keep stack data valid */
```

The library returns `USBPD_OK`, but what gets sent on the CC line is an invalid **Get_Battery_Status** message. `Size == 0` in the extended header.

When I set `uint8_t data = 1`, then `Size = 1` in the extended header, which is better.. however, the `Battery Status Ref` is set to `0x40`, which is invalid.

# Info
* USB PD 3.0 V2.0 20190828: Section 6.5.4
* This repo (v4.2.0 USB PD release)

# Screenshots
These were taken from Total Phase's USB PD sniffer + their DataCenter software
1. data = 0. Note the "Size" area in pink is 0, but should be '1'
![dataIsSetTo0](https://github.com/STMicroelectronics/stm32-mw-usbpd-core/assets/741701/bd62ad1c-742e-4fe4-9dc3-2e1cef688165)

2. data = 1. Note the "Size" area is '1', but the `Battery Status Ref` (incorrectly named to `Battery Cap Ref`), is incorrectly set to `0x40`, not '1'. See "Data" in Green
![dataIsSetTo1](https://github.com/STMicroelectronics/stm32-mw-usbpd-core/assets/741701/9a56e4f7-5027-46bf-bfb2-2044addff942)

This is a critical piece for our application. Please let me know if you'd like any more information.

**Comment by HFISTM** [comment]
Hello,

Thank you for this report. We will get back to you as soon as we analyze it further. This may take some time. Thank you for your comprehension.

With regards,

**Comment by HFISTM** [comment]
ST Internal Reference: 156169

**Comment by davedesro** [comment]
Hi @HFISTM thank you.

I have an update: the Total Phase PD Sniffer itself is reporting bad data (there's a bug in the PD Sniffer data parsing).

Still, for Battery_Status command, the library is not giving any notification that one was received (I confirmed at the HAL that the return data is valid)

Also, when the port-partner returns Battery_Status command, the library _always_ issues a SOFT_RESET back to the port-partner.

So, in summary, the issue as I see things now are
1) Upon Battery_Status inbound command, there is no Battery_Status notification (No call to `USBPD_DPM_SetDataInfo()`)
2) Upon Battery_Status inbound command, SOFT_RESET is always issued

**Comment by davedesro** [comment]
Hi @HFISTM Update. This issue is seen with a Samsung S20.

I get a tag of 15 (`USBPD_CORE_BATTERY_STATUS`) in `USBPD_DPM_SetDataInfo()` for an Anker power brick.

**Comment by davedesro** [comment]
@HFISTM one more observation: when in `USBPD_DPM_SetDataInfo()`, if it copy the `Battery_Status` data in `Ptr-2`, it seems to work. (see image.. "Ptr-2")
Is it possible the library is incorrectly looking at +2 byte offset of the `Battery_Status` response?

![image](https://github.com/STMicroelectronics/stm32-mw-usbpd-core/assets/741701/3a3c19dd-123d-4d4b-a762-784072c8655a)

**Comment by HFISTM** [comment]
Hello @davedesro , to be sure to understand, what is your Datatype_BatteryStatusInfo field ? SetDataInfo(USBPD_CORE_BATTERY_STATUS) is supposed to write Ptr direclty (no offset) to DPM_BatteryStatus field, of type USBPD_BSDO_TypeDef which is defined in the PD spec. 
Pleaase refer to [this application](https://github.com/STMicroelectronics/STM32CubeG0/blob/ae31d181b3244190d7d5bc0d91e66a82ce4270ce/Projects/STM32G081B-EVAL/Demonstrations/DemoUCPD/Src/usbpd_dpm_user.c#L856) as exemple.

Also could you confirm your issue still present on latest library (v5.0.0) ? I noticed you were on v4.2.0.
Thanks

**Comment by davedesro** [comment]
Hi @HFISTM I updated to 5.0.0 and see the same issue... I need to subtract by 2 (`Ptr-2`).

Thank you for the link to the application. I believe there may be an issue with the code.

![image](https://github.com/STMicroelectronics/stm32-mw-usbpd-core/assets/741701/94551ab1-7596-4d38-8510-4865069e1d25)

https://github.com/STMicroelectronics/STM32CubeG0/blob/master/Projects/STM32G081B-EVAL/Demonstrations/DemoUCPD/Src/usbpd_dpm_user.c#L856

The code is copying only 1 byte (Ptr[0]) to a `uint32_t`.  Shouldn't it copy 4 bytes? This is incorrect, but it might also hide the `Ptr-2` issue

**Comment by davedesro** [comment]
@HFISTM There may be another issue with 5.0.0, 

When the library is in UFP+SRC mode, the functions `USBPD_PE_Request_CtrlMessage()` `USBPD_PE_SendExtendedMessage()` do not work. SVDM responses still work, however.

All other combinations of UFP/DFP + SNK/SRC work fine. I'm able to call ctrl messages.

Shall I create another ticket for this issue, since it seems like a regression?

In the meantime, I will continue to use v4.2.0

**Comment by HFISTM** [comment]
Hello @davedesro, 
Thank you for your feedback. **For any new issue, please open another ticket with details.**

Regarding the battery status, you are right about the copy of only 1 byte instead of 4.
Could you please try with the following code ?

```
  case USBPD_CORE_BATTERY_STATUS:
    {
      uint8_t* battery_status;
      battery_status = (uint8_t*)&DPM_Ports[PortNum].DPM_BatteryStatus.d32;
      memcpy(battery_status, Ptr, Size);
    }
```

This might also fix the "-2" issue, as the previous code was only taking the first byte of the message, which was not the actual BatteryInfo response. Please refer to struct USBPD_BSDO_TypeDef: batteryInfo field is the 3rd of the 4 bytes in this struct. As you were only copying the first byte, substracting 2 to the pointer directly copied the 3rd byte, meaning the BatteryInfo you were looking for.

With this fix you should be able to access the BatteryInfo field through the USBPD_BSDO_TypeDef struct directly as it will now be fully copied by SetDataInfo function.

Thank you & regards

**Comment by mroszkobt** [comment]
I'm playing around with the PD stack recently on a STM32G0B1.

I request ManufacturerInfo,BatteryStatus and BatteryCapabilities (with 500ms) between each request.

The callback for USBPD_CORE_BATTERY_STATUS is consistently wrong. 

It has nothing to do with the lack of memcpy above.

```

    case USBPD_CORE_BATTERY_STATUS:
      if (Size <= sizeof(USBPD_BSDO_TypeDef))
      {
        memcpy((uint8_t*)&DPM_Ports[PortNum].DPM_BatteryStatus.d32, Ptr, Size);
        
        if(DPM_Ports[PortNum].DPM_BatteryStatus.b.BatteryPresentCapacity != UINT16_MAX)
        {
          DPM_Ports[PortNum].DPM_BatteryStatusValid = true;
          if( DPM_Ports[PortNum].DPM_BatteryCapValid )
          {
            // precalculate
            uint32_t temp = DPM_Ports[PortNum].DPM_BatteryStatus.b.BatteryPresentCapacity;
            temp *= 100;
            temp /= DPM_Ports[PortNum].DPM_BatteryCapability.BatteryLastFullChargeCapa;
            DPM_Ports[PortNum].DPM_BatteryRSOC = (uint16_t)temp;
          }
        }
      }
      break;
```


In my situation, when I view the variable data in the callback, the Size is correctly "4". 

But the *Ptr data is still pointing at the last extended message response which was the manufacturerinfo response!

**Comment by mroszkobt** [comment]
> [@HFISTM](https://github.com/HFISTM) one more observation: when in `USBPD_DPM_SetDataInfo()`, if it copy the `Battery_Status` data in `Ptr-2`, it seems to work. (see image.. "Ptr-2") Is it possible the library is incorrectly looking at +2 byte offset of the `Battery_Status` response?
> 
> ![image](https://github.com/STMicroelectronics/stm32-mw-usbpd-core/assets/741701/3a3c19dd-123d-4d4b-a762-784072c8655a)

Thanks for finding the data sits at -2, saved me a hassle ;)
